src/util/network_namespace.cc

```diff
@@ -0,0 +1,168 @@
+/* -*-mode:c++; tab-width: 4; indent-tabs-mode: nil; c-basic-offset: 4 -*- */
+#include "network_namespace.hh"
+
+#include <stdio.h>
+#include <string.h>
+#include <unistd.h>
+#include <fcntl.h>
+#include <dirent.h>
+#include <errno.h>
+#include <sys/mount.h>
+#include <sys/stat.h>
+#include <sched.h>
+
+#include "config.h"
+#include "system_runner.hh"
+#include "exception.hh"
+#include "file_descriptor.hh"
+
+using namespace std;
+
+NetworkNamespace::NetworkNamespace( const std::string & name )
+    : name_(name), has_own_resolvconf_(false), has_entered_(false)
+{
+    /* create a new network namespace */
+    run( { IP, "netns", "add", name_ } );
+}
+
+
+NetworkNamespace::~NetworkNamespace()
+{
+    if ( has_entered_ ) {
+
+        char etc_netns_path[PATH_MAX];
+        char netns_name[PATH_MAX];
+        char etc_name[PATH_MAX];
+        struct dirent *entry;
+        DIR *dir;
+
+        snprintf( etc_netns_path, sizeof( etc_netns_path ), "%s/%s", NETNS_ETC_DIR.c_str(), name_.c_str() );
+        dir = opendir( etc_netns_path );
+        if (!dir )
+            return;
+
+        while ( ( entry = readdir(dir) ) != NULL ) {
+
+            if ( strcmp(entry->d_name, ".") == 0 )
+                continue;
+            if ( strcmp(entry->d_name, "..") == 0 )
+                continue;
+
+            snprintf(netns_name, sizeof(netns_name), "%s/%s", etc_netns_path, entry->d_name);
+            snprintf(etc_name, sizeof(etc_name), "/etc/%s", entry->d_name);
+
+            if ( umount( etc_name ) < 0 ) {
+                throw runtime_error( string("Unmounting... ") + netns_name + " -> " + etc_name + " failed: " + strerror(errno) + "\n" );
```

Comment:
This line causes the compile to fail on g++ 6.2. (Can't throw from a destructor.)

`network_namespace.cc: In destructor ‘NetworkNamespace::~NetworkNamespace()’:
network_namespace.cc:55:135: error: throw will always call terminate() [-Werror=terminate]
 . ") + netns_name + " -> " + etc_name + " failed: " + strerror(errno) + "\n" );
                                                                              ^
network_namespace.cc:55:135: note: in C++11 destructors default to noexcept
`